_source/_posts/2022-06-01-oauth-public-client-identity.md

```diff
@@ -0,0 +1,111 @@
+
+{% img blog/oauth-public-client-identity/authorization-screen.png alt:"An OAuth consent screen asking the user whether it is okay for the app to see their email address and files" width:"600" %}{: .center-image }
+
+This accomplishes basically nothing if you do this to your own account on your own device, but is potentially more interesting if you can accomplish this on someone else's device.
+
+A typical sequence when logging in to an application is to click the "log in" button, enter your credentials, be prompted with a consent screen, and be redirected back to the application.  
+
+{% img blog/oauth-public-client-identity/mobile-sequence.png alt:"Sequence of steps logging in to an application" width:"600" %}{: .center-image }
+
+The authorization server may choose to skip the credential prompt if the user is already logged in. The authorization server may also choose to skip the consent screen if the user has already consented to this application before or for first-party apps. In that case, the redirect would be immediate, which is good for user experience, but bad if client impersonation is possible. Again, this isn't a concern for confidential clients since client impersonation isn't a concern if the client can authenticate. But for public clients, since the OAuth flow is done with only public information, client impersonation is definitely possible. That said, the original OAuth 2.0 specification actually calls this out explicitly in the [Security Considerations Section 10.2](https://datatracker.ietf.org/doc/html/rfc6749#section-10.2). The recommended mitigation from the original spec is to require redirect URL registration and to avoid skipping the consent screen. This gives the end user a chance to spot any attempts at tricking them into authorizing a malicious application.
+
+## Gaining privileged API access
+
+Some services may restrict certain APIs for use only by specific clients. If the service has some APIs that should only be called from a particular mobile app, the service may want to reject requests made from other apps that are accessing other parts of the system. Additionally, some services may set different API rate limits for different clients depending on the trust level of the developer or whether the client is a first-party or third-party client. 
+
+In order to do this, the API needs to know which client an access token was issued to. Typically the authorization server will embed the application's client_id in one of the claims of the access token so that it's accessible to the API when it validates the token. Of course this information is only as good as the authorization server's confidence that it was really that app making the token request. Again, if the client can authenticate itself, then this is reasonably reliable. But for public clients, there is much less certainty.
```

Comment:
```suggestion
In order to do this, the API needs to know which client an access token was issued to. Typically, the authorization server will embed the application's client_id in one of the claims of the access token. In this way, the client_id is accessible to the API when it validates the token. Of course, this information is only as good as the authorization server's confidence that it was really that app making the token request. Again, if the client can authenticate itself, then this is reasonably reliable. But for public clients, there is much less certainty.
```
This sentence was tripping me up when re-reading, so I made it 2 simpler sentences